### ISSUE TITLE
Possible regression in 3.9-beta when destructuring from a union including undefined

### ISSUE THREAD

**Issue by shawnmcknight** [microsoft/TypeScript#38110]
**TypeScript Version:**  Nightly

**Search Terms:**
`undefined`, `destructure`, `destructuring`

**Code**

```ts
interface Foo {
  foo1: string
}

let foo: Foo | undefined;
const { foo1 } = foo; // foo1: string
```

**Expected behavior:**
As in 3.8, the `const { foo1 } = foo;` should emit error `Property 'foo1' does not exist on type 'Foo | undefined'.(2339)`

**Actual behavior:**
No error emitted.  Type of `foo1` is inferred as `string`.

**Playground Link:** 
https://www.typescriptlang.org/play/?ts=Nightly#code/JYOwLgpgTgZghgYwgAgGIHt3IN4ChnIyYCMAXMgM5hSgDmuAvrrgDYRiGbkZYA+yAVxAATCDFARhAblwJ0IKjk7piyBsgC8yqcgD0u5WUrU6QA

**Related Issues:** 
None found
